Define a^s for matrices (#3805)
modelica/ModelicaSpecification: chapters/arrays.tex

\subsection{Scalar Exponentiation of Matrices}\label{scalar-exponentiation-of-square-matrices-of-numeric-elements}\label{scalar-exponentiation-of-matrices}

> Exponentiation \lstinline!a ^ s! is defined if \lstinline!a! is a square matrix of a subtype of \lstinline!Real!, and \lstinline!s! is a scalar subtype of \lstinline!Integer! with $\text{\lstinline!s!} \geq 0$.

Reviewer: According to yesterday's meeting:

```suggestion
Exponentiation \lstinline!a ^ s! is defined if \lstinline!a! is a square numeric matrix, and \lstinline!s! is a scalar subtype of \lstinline!Integer! with $\text{\lstinline!s!} \geq 0$.
The result is a matrix of the same type as \lstinline!a!.
The exponentiation is defined mathematically as repeated multiplication, with an identity matrix as the result for $\text{\lstinline!s!} = 0$, e.g.:
```

(Stupid Github doesn't allow the suggestion to also replace the two lines below because of presence of earlier conversations which have already been resolved.)

Reviewer: Resolving outdated conversation.

> The result is a \lstinline!Real! matrix.

Reviewer: Workaround for issue with Gitbug suggestion mentioned above:

```suggestion
```

> The exponentiation is defined mathematically as repeated multiplication, with a \lstinline!Real! identity matrix as the result for $\text{\lstinline!s!} = 0$, e.g.:

Reviewer: Workaround for issue with Github suggestion mentioned above:

```suggestion
```

\subsection{Scalar Exponentiation of Matrices}\label{scalar-exponentiation-of-square-matrices-of-numeric-elements}\label{scalar-exponentiation-of-matrices}

> Exponentiation \lstinline!a ^ s! is defined if \lstinline!a! is a square matrix of a subtype of \lstinline!Real!, and \lstinline!s! is a scalar as a subtype of \lstinline!Integer! with $\text{\lstinline!s!} \geq 0$.

Reviewer: There's an old typo we should fix too.  Further, to keep this aligned with how other operators are defined, we should also specify explicitly that the result is a `Real` matrix.

```suggestion
Exponentiation \lstinline!a ^ s! is defined if \lstinline!a! is a square matrix of a subtype of \lstinline!Real!, and \lstinline!s! is a scalar subtype of \lstinline!Integer! with $\text{\lstinline!s!} \geq 0$.
The result is a \lstinline!Real! matrix.
```

> The exponentiation is done by repeated multiplication, e.g.:

Reviewer: Actually performing repeated multiplication sounds like a bad idea to me.  Perhaps we should also make an effort to avoid specifying _how_ to compute the result, in order to give room for implementations to compute something closer to a correctly rounded result?

Simply saying _defined as_ instead of _done by_ would be a step in the right direction:

```suggestion
The exponentiation is defined as repeated multiplication, e.g.:
```

Reviewer: Well, when I head "done by repeated multiplication", I only hear two possible interpretations; repeated multiplication from left, and repeated multiplication from right.  This also points at a danger with using some sort of Modelica pseudo code in the semi-example included in the definition, as the Modelica syntax actually comes with a particular choice of operator associativity.

> The exponentiation is defined as repeated multiplication, e.g.:

Reviewer: Would it also help to throw in _mathematically_ to reduce risk of interpreting the examples as some sort of reference implementation?

```suggestion
The exponentiation is defined mathematically as repeated multiplication, with a \lstinline!Real! identity matrix as the result for $\text{\lstinline!s!} = 0$, e.g.:
```

a^3 = a * a * a;
a^0 = identity(size(a, 1)) * 1.0;
assert(size(a, 1) == size(a, 2), "Matrix must be square");

> a^1 = a;

Author: I re-ordered them, replaced the assert by a^2, and replaced the zero-case by text.

> a^1 = a;

Reviewer: Continued in another conversation.  Resolving.
